### PR TITLE
Signaling Implementation

### PR DESCRIPTION
**Pull request by ActoryOu**
Issue #, if available:
Signaling implementation.

Description of changes:
Make Signaling standalone and separate out the APIs.
Couple with standard C library and [coreJSON](https://github.com/FreeRTOS/coreJSON) only.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by aggarg — DISMISSED** [review]

File: README.md [hunk 1]
```diff
@@ -1,17 +1,36 @@
...
-
-* Change the title in this README
-* Edit your repository description on GitHub
-
-## Security
-
-See [CONTRIBUTING](CONTRIBUTING.md#security-issue-notifications) for more information.
+## amazon-kinesis-video-streams-signaling
+
+The goal of the Signaling library is to enable control plane communication
```
Comment (unicornss) [inline README.md]: is this just for control plane?

File: README.md [hunk 2]
```diff
@@ -1,17 +1,36 @@
...
+- Note that this library has a dependency on [coreJSON](https://github.com/FreeRTOS/coreJSON).
+
+## Using the library
+
+1. Use the `Signaling_Construct*` APIs to construct requests to send to the
+   Signaling Service.
+    - Ensure authenticate to sign the constructed requests using the Signature
+      Version 4 (SigV4) [authentication flow](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html)
+      before sending them.
+3. Use the `Signaling_Parse*` APIs to parse the responses from the Signaling
```
Comment (unicornss) [inline README.md]: /nit numbering ?

File: source/include/signaling_data_types.h [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+/* Standard includes. */
+#include <stdint.h>
+#include <stddef.h>
+
+/*-----------------------------------------------------------*/
+
+/**
+ * Maximum number of URIs for an ICE server.
+ */
+#define SIGNALING_ICE_SERVER_MAX_URIS ( 4 )
```
Comment (unicornss) [inline source/include/signaling_data_types.h:13]: is this the list size max?

File: source/include/signaling_data_types.h [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+    SIGNALING_PROTOCOL_MAX = 0xFF,
+} SignalingProtocol_t;
+
+/**
+ * @ingroup signaling_enum_types
+ * @brief Type of signaling channel.
+ *        Refer to https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_ChannelInfo.html#KinesisVideo-Type-ChannelInfo-ChannelType for more detail.
+ */
+/* We cannot name it SignalingChannelType_t in order to avoid naming conflict
+ * with a data type in the existing KVS SDK. */
```
Comment (unicornss) [inline source/include/signaling_data_types.h:83]: signaling channel type?

File: source/include/signaling_data_types.h [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+    SIGNALING_TYPE_MESSAGE_GO_AWAY,
+    SIGNALING_TYPE_MESSAGE_RECONNECT_ICE_SERVER,
+    SIGNALING_TYPE_MESSAGE_STATUS_RESPONSE,
+} SignalingTypeMessage_t;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @ingroup signaling_enum_types
+ * @brief Basic format of the signaling request.
```
Comment (unicornss) [inline source/include/signaling_data_types.h:124]: could you add details on the structure attributes? how about port or protocol?

File: source/include/signaling_data_types.h [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+ */
+typedef struct WssRecvMessage
+{
+    const char * pSenderClientId;
+    size_t senderClientIdLength;
+    SignalingTypeMessage_t messageType;
+    const char * pBase64EncodedPayload;
+    size_t base64EncodedPayloadLength;
+    WssStatusResponse_t statusResponse;
+    SignalingIceServer_t * pIceServers; /* The caller needs to supply memory. */
```
Comment (unicornss) [inline source/include/signaling_data_types.h]: can you make it consistent across all structs?

File: source/signaling_api.c [hunk 7]
```diff
@@ -0,0 +1,1810 @@
...
+{
+    SignalingResult_t result = SIGNALING_RESULT_OK;
+    int snprintfRetVal = 0;
+    size_t i, j;
+    size_t remainingLength = *pRemainingLength;
+    size_t currentIndex = *pCurrentIndex;
+
+    snprintfRetVal = snprintf( &( pBuffer[ currentIndex ] ),
+                               remainingLength,
+                               ","                                                                                                                                            \
```
Comment (hassanctech) [inline source/signaling_api.c]: Stray `\` on line?

File: README.md (hunk 1 above)
Comment (aggarg) [reply README.md]: Corrected.

File: README.md (hunk 2 above)
Comment (aggarg) [reply README.md]: Fixed.

File: source/include/signaling_data_types.h (hunk 3 above)
Comment (aggarg) [reply source/include/signaling_data_types.h:13]: This is maximum number of URIs for an ICE server and it is currently fixed. The current SDK seems to support one - https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Ice/IceUtils.h#L56.

File: source/include/signaling_data_types.h (hunk 5 above)
Comment (aggarg) [reply source/include/signaling_data_types.h:124]: We plan to do documentation for all the libraries later before we go GA.

File: source/include/signaling_data_types.h (hunk 6 above)
Comment (aggarg) [reply source/include/signaling_data_types.h]: Assuming that you are talking about memory related comment, it is already consistent across all the structs. I guess this comment is uncalled for - I have removed it.

File: source/signaling_api.c (hunk 7 above)
Comment (aggarg) [reply source/signaling_api.c]: Good catch! Fixed.

File: source/include/signaling_data_types.h (hunk 4 above)
Comment (aggarg) [reply source/include/signaling_data_types.h:83]: This comment is not clear to me. Would you please elaborate.

File: .gitmodules [hunk 8]
```diff
@@ -0,0 +1,4 @@
+[submodule "source/dependency/coreJSON"]
+	path = source/dependency/coreJSON
+	url = https://github.com/FreeRTOS/coreJSON
+	update = none
```
Comment (sirknightj) [inline .gitmodules:4]: Curious why not lock in a specific commit?

File: source/include/signaling_data_types.h (hunk 3 above)
Comment (unicornss) [reply source/include/signaling_data_types.h:13]: Service response `IceServerList` should contain two items with `3 URIs` (now) in each item.

File: README.md (hunk 1 above)
Comment (sirknightj) [reply README.md]: the TURN servers (https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_Operations_Amazon_Kinesis_Video_Signaling_Channels.html) and the Websocket APIs belong to data plane

File: README.md [hunk 9]
```diff
@@ -1,17 +1,36 @@
...
+    - Ensure authenticate to sign the constructed requests using the Signature
+      Version 4 (SigV4) [authentication flow](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html)
+      before sending them.
+2. Use the `Signaling_Parse*` APIs to parse the responses from the Signaling
+   Service.
+3. Use `Signaling_ConstructWssMessage` and `Signaling_ParseWssRecvMessage` APIs
+   to communicate with the WSS endpoint.
+    - Ensure to authenticate and sign the constructed messages using the
+      Signature Version 4 (SigV4) [authentication flow](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html)
+      before sending them.
```
Comment (sirknightj) [inline README.md:32]: Any build or consumption instructions?

File: source/include/signaling_data_types.h [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+} SignalingTypeChannel_t;
+
+/**
+ * @ingroup signaling_enum_types
+ * @brief Role of current signaling request.
+ *        Refer to https://docs.aws.amazon.com/kinesisvideostreams-webrtc-dg/latest/devguide/kvswebrtc-how-it-works.html for more detail.
+ */
+typedef enum SignalingRole
+{
+    SIGNALING_ROLE_NONE = 0,
```
Comment (sirknightj) [inline source/include/signaling_data_types.h:97]: What is the purpose of SIGNALING_ROLE_NONE? Is it used somewhere?

File: source/signaling_api.c [hunk 11]
```diff
@@ -0,0 +1,1810 @@
...
+/* API includes. */
+#include "signaling_api.h"
+
+/* CoreJSON includes. */
+#include "core_json.h"
+
+/**
+ * Helper macro to check if the AWS region is China region.
+ */
+#define SIGNALING_IS_CHINA_REGION( pAwsRegion )     \
```
Comment (sirknightj) [inline source/signaling_api.c:15]: What about other regions? 

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/3e519b7670e39031375d227f983ad2cde888078e/tst/ProducerCallbackProviderApiTest.cpp#L66-L88

File: source/signaling_api.c [hunk 12]
```diff
@@ -0,0 +1,1810 @@
...
+    case SIGNALING_TYPE_MESSAGE_SDP_ANSWER:
+        ret = "SDP_ANSWER";
+        break;
+
+    case SIGNALING_TYPE_MESSAGE_ICE_CANDIDATE:
+        ret = "ICE_CANDIDATE";
+        break;
+
+    default:
+        ret = "UNKOWN";
```
Comment (sirknightj) [inline source/signaling_api.c]: Spelling

File: source/signaling_api.c [hunk 13]
```diff
@@ -0,0 +1,1810 @@
...
+
+    if( ( pAwsRegion == NULL ) ||
+        ( pAwsRegion->pAwsRegion == NULL ) ||
+        ( pChannelArn == NULL ) ||
+        ( pRequestBuffer == NULL ) ||
+        ( pRequestBuffer->pUrl == NULL ) ||
+        ( pRequestBuffer->pBody == NULL ) ||
+        ( pChannelArn->pChannelArn == NULL ) )
+    {
+        result = SIGNALING_RESULT_BAD_PARAM;
```
Comment (sirknightj) [inline source/signaling_api.c:455]: Can we also add empty string checks.

File: source/signaling_api.c [hunk 14]
```diff
@@ -0,0 +1,1810 @@
...
+    }
+
+    if( result == SIGNALING_RESULT_OK )
+    {
+        if( SIGNALING_IS_CHINA_REGION( pAwsRegion ) )
+        {
+            snprintfRetVal = snprintf( pRequestBuffer->pUrl,
```
Comment (sirknightj) [inline source/signaling_api.c:462]: Should add other cases?

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/pull/430/files

File: source/signaling_api.c (hunk 13 above)
Comment (unicornss) [reply source/signaling_api.c:455]: pChannelArn->pChannelArn == NULL

File: source/include/signaling_data_types.h (hunk 3 above)
Comment (ActoryOu) [reply source/include/signaling_data_types.h:13]: Hi @unicornss,
Thanks for your review! This is a nice one!

We refer to [IceConfigInfo](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/include/com/amazonaws/kinesis/video/webrtcclient/Include.h#L1396), the [MAX_ICE_CONFIG_URI_COUNT](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/include/com/amazonaws/kinesis/video/webrtcclient/Include.h#L615) in it is defined to 4. From the past log, I can only see 3 URIs in a single Ice server, but I thought there might be some reasons to define it to 4 in current SDK. We don't want to change any behavior comparing to current implementation. If it's sure that there are at most 3 URIs, I'm totally fine to update it.

BTW, I can't find the definition in [KVS website](https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_signaling_GetIceServerConfig.html). Could you help guide me to the right place?

Thank you.

File: source/include/signaling_data_types.h (hunk 4 above)
Comment (ActoryOu) [reply source/include/signaling_data_types.h:83]: The definitions here are conflicted with [here in existing KVS SDK](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/4f1d3add4c0f88212fe6646898e5aba140ecbb3c/src/include/com/amazonaws/kinesis/video/webrtcclient/Include.h#L925-L928).

File: source/include/signaling_data_types.h (hunk 10 above)
Comment (ActoryOu) [reply source/include/signaling_data_types.h:97]: In my experience, 0 is too easy to be misused, which means user might set role to 0 by accident. That's why I set 0 as unused or none or unknown in most of the enums in this component.

File: .gitmodules (hunk 8 above)
Comment (ActoryOu) [reply .gitmodules:4]: In git submodule, the commit of this repo is bound automactically. Take [this folder](https://github.com/ActoryOu/amazon-kinesis-video-streams-signaling-public/tree/aggarg_suggestions/source/dependency) as reference. It shows that [dc1ab91](https://github.com/FreeRTOS/coreJSON/tree/dc1ab9130a1fb99b801a2a1fa8e9f42239f752be) is the exact commit of coreJSON bound in signaling repo.

File: README.md (hunk 9 above)
Comment (ActoryOu) [reply README.md:32]: Actually, this part should be done in application layer. We'd provide references (e.g. [in existing SDK](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1950)).

File: source/signaling_api.c (hunk 11 above)
Comment (ActoryOu) [reply source/signaling_api.c:15]: Thanks for sharing! I refer to [this implementation](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/4f1d3add4c0f88212fe6646898e5aba140ecbb3c/src/source/Signaling/ChannelInfo.c#L183). Could you help me understand if we need more special rule for other region? Is it necessary to apply that into WebRTC? In my understanding, we're not using curl in WebRTC flow.

File: source/signaling_api.c (hunk 12 above)
Comment (ActoryOu) [reply source/signaling_api.c]: Thanks! Updated!

File: source/signaling_api.c (hunk 13 above)
Comment (ActoryOu) [reply source/signaling_api.c:455]: That's checked at the end.

File: source/signaling_api.c (hunk 14 above)
Comment (ActoryOu) [reply source/signaling_api.c:462]: Again, could you help me understand if it is necessary to apply that into WebRTC? In my understanding, we're not using curl in WebRTC flow.

File: .gitmodules (hunk 8 above)
Comment (sirknightj) [reply .gitmodules:4]: Makes sense, thanks for the clarification!

File: source/include/signaling_data_types.h (hunk 3 above)
Comment (sirknightj) [reply source/include/signaling_data_types.h:13]: 337520443-4994fdfb-37d0-4f4f-910b-201327c7f5a3


https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_signaling_IceServer.html#KinesisVideo-Type-signaling_IceServer-Uris

Doesn't seem to be any exact URL list length, just max 256 chars.

File: source/signaling_api.c (hunk 11 above)
Comment (hassanctech) [reply source/signaling_api.c:15]: This is correct we do not need special rule for other regions for webrtc.  If they are needed in the future they can be added as requested.

Comment (sirknightj) [reply source/signaling_api.c:15]: I reached out to confirm the above, and what you mentioned is accurate .

File: source/signaling_api.c (hunk 14 above)
Comment (sirknightj) [reply source/signaling_api.c:462]: Not on the current roadmap so not needed in this pr

File: README.md [hunk 15]
```diff
@@ -1,17 +1,36 @@
...
+signaling messages. This library provides APIs to construct requests to send to
+the Signaling Service and to parse responses from the Signaling Service.
+
+- Note that this library has a dependency on [coreJSON](https://github.com/FreeRTOS/coreJSON).
+
+## Using the library
+
+1. Use the `Signaling_Construct*` APIs to construct requests to send to the
+   Signaling Service.
+    - Ensure authenticate to sign the constructed requests using the Signature
```
Comment (sirknightj) [inline README.md]: Ensure to authenticate and sign

File: source/include/signaling_data_types.h [hunk 16]
```diff
@@ -0,0 +1,381 @@
...
+{
+    const char * pSenderClientId;
+    size_t senderClientIdLength;
+    SignalingTypeMessage_t messageType;
+    const char * pBase64EncodedPayload;
+    size_t base64EncodedPayloadLength;
+    WssStatusResponse_t statusResponse;
+    SignalingIceServer_t * pIceServers;
+    size_t numIceServers;
+} WssRecvMessage_t;
```
Comment (sirknightj) [inline source/include/signaling_data_types.h:374]: I think you should clarify why `pIceServers` and `numIceServers` in this struct.

I don't see it in https://docs.aws.amazon.com/kinesisvideostreams-webrtc-dg/latest/devguide/kvswebrtc-websocket-apis-7.html

Seems that this data type is a union of the above plus the geticeserverconfig api.. which is not sent via secure websocket (https)

File: README.md (hunk 15 above)
Comment (ActoryOu) [reply README.md]: Updated

File: source/include/signaling_data_types.h (hunk 16 above)
Comment (ActoryOu) [reply source/include/signaling_data_types.h:374]: Updated comments.

**Review by hassanctech — DISMISSED** [review]
lgtm

**Review by aggarg — APPROVED** [review]

File: source/include/signaling_data_types.h [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+
+/**
+ * @ingroup signaling_enum_types
+ * @brief Protocol type of endpoints while parsing get signaling channel endpoint response.
+ *        Refer to https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_GetSignalingChannelEndpoint.html for more detail.
+ */
+typedef enum SignalingProtocol
+{
+    SIGNALING_PROTOCOL_NONE = 0,
+    SIGNALING_PROTOCOL_WEBSOCKET_SECURE = 1,
```
Comment (sirknightj) [inline source/include/signaling_data_types.h:71]: Want to check, that the name should be `WEBSOCKET_SECURE` instead of `WSS`?

Comment (sirknightj) [reply source/include/signaling_data_types.h:71]: https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/4f1d3add4c0f88212fe6646898e5aba140ecbb3c/src/source/Signaling/LwsApiCalls.h#L25-L26

> // Protocol indexes
#define PROTOCOL_INDEX_HTTPS 0
#define PROTOCOL_INDEX_WSS   1

**Review by sirknightj — APPROVED** [review]

File: source/include/signaling_data_types.h (hunk 17 above)
Comment (ActoryOu) [reply source/include/signaling_data_types.h:71]: Yeah, would it be any concern of changing this name?

**Pull request merged**